File: src/pick_place_controller.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/joint_state.hpp>
#include <trajectory_msgs/msg/joint_trajectory.hpp>
#include <trajectory_msgs/msg/joint_trajectory_point.hpp>
#include <std_msgs/msg/float64_multi_array.hpp>
#include <vector>
#include <chrono>

class PickPlaceController : public rclcpp::Node
{
public:
    PickPlaceController() : Node("pick_place_controller")
    {
        // Publishers
        joint_trajectory_pub_ = this->create_publisher<trajectory_msgs::msg::JointTrajectory>(
            "/pick_place_robot/joint_trajectory_controller/joint_trajectory", 10);
        
        gripper_pub_ = this->create_publisher<std_msgs::msg::Float64MultiArray>(
            "/pick_place_robot/gripper_controller/commands", 10);

        // Subscriber
        joint_state_sub_ = this->create_subscription<sensor_msgs::msg::JointState>(
            "/pick_place_robot/joint_states", 10,
            std::bind(&PickPlaceController::joint_state_callback, this, std::placeholders::_1));

        // Initialize joint names
        joint_names_ = {"joint1", "joint2", "joint3"};
        current_joint_positions_.resize(3, 0.0);
        
        // Timer for pick and place sequence
        timer_ = this->create_wall_timer(
            std::chrono::seconds(2),
            std::bind(&PickPlaceController::execute_pick_place_sequence, this));

        RCLCPP_INFO(this->get_logger(), "Pick and Place Controller initialized");
        
        // Initialize sequence
        sequence_step_ = 0;
        gripper_closed_ = false;
    }

private:
    void joint_state_callback(const sensor_msgs::msg::JointState::SharedPtr msg)
    {
        for (size_t i = 0; i < msg->name.size(); ++i) {
            for (size_t j = 0; j < joint_names_.size(); ++j) {
                if (msg->name[i] == joint_names_[j]) {
                    current_joint_positions_[j] = msg->position[i];
                    break;
                }
            }
        }
    }

    void move_to_joint_positions(const std::vector<double>& target_positions, double duration = 3.0)
    {
        auto trajectory_msg = trajectory_msgs::msg::JointTrajectory();
        trajectory_msg.header.stamp = this->now();
        trajectory_msg.joint_names = joint_names_;

        auto point = trajectory_msgs::msg::JointTrajectoryPoint();
        point.positions = target_positions;
        point.time_from_start = rclcpp::Duration::from_seconds(duration);

        trajectory_msg.points.push_back(point);
        joint_trajectory_pub_->publish(trajectory_msg);
        
        RCLCPP_INFO(this->get_logger(), "Moving to joint positions: [%.2f, %.2f, %.2f]",
                    target_positions[0], target_positions[1], target_positions[2]);
    }

    void control_gripper(bool close)
    {
        auto gripper_msg = std_msgs::msg::Float64MultiArray();
        if (close) {
            gripper_msg.data = {0.04, 0.04}; // Close gripper
        } else {
            gripper_msg.data = {0.0, 0.0};   // Open gripper
        }
        gripper_pub_->publish(gripper_msg);
        gripper_closed_ = close;
        
        RCLCPP_INFO(this->get_logger(), "Gripper %s", close ? "closed" : "opened");
    }

    void execute_pick_place_sequence()
    {
        switch (sequence_step_) {
            case 0:
                RCLCPP_INFO(this->get_logger(), "Step 1: Moving to home position");
                move_to_joint_positions({0.0, 0.0, 0.0});
                control_gripper(false);
                break;
                
            case 1:
                RCLCPP_INFO(this->get_logger(), "Step 2: Moving to pick position (red box)");
                // Position above red box at (0.5, 0.2, 1.05)
                move_to_joint_positions({0.4, -0.3, 0.5});
                break;
                
            case 2:
                RCLCPP_INFO(this->get_logger(), "Step 3: Lowering to pick");
                move_to_joint_positions({0.4, -0.1, 0.3});
                break;
                
            case 3:
                RCLCPP_INFO(this->get_logger(), "Step 4: Closing gripper");
                control_gripper(true);
                break;
                
            case 4:
                RCLCPP_INFO(this->get_logger(), "Step 5: Lifting object");
                move_to_joint_positions({0.4, -0.3, 0.5});
                break;
                
            case 5:
                RCLCPP_INFO(this->get_logger(), "Step 6: Moving to place position");
                // Position above target area at (-0.5, 0, 1.025)
                move_to_joint_positions({-0.8, -0.3, 0.5});
                break;
                
            case 6:
                RCLCPP_INFO(this->get_logger(), "Step 7: Lowering to place");
                move_to_joint_positions({-0.8, -0.1, 0.3});
                break;
                
            case 7:
                RCLCPP_INFO(this->get_logger(), "Step 8: Opening gripper");
                control_gripper(false);
                break;
                
            case 8:
                RCLCPP_INFO(this->get_logger(), "Step 9: Retracting");
                move_to_joint_positions({-0.8, -0.3, 0.5});
                break;
                
            case 9:
                RCLCPP_INFO(this->get_logger(), "Step 10: Returning to home");
                move_to_joint_positions({0.0, 0.0, 0.0});
                break;
                
            case 10:
                RCLCPP_INFO(this->get_logger(), "Pick and place sequence completed!");
                // Reset sequence to repeat
                sequence_step_ = -1;
                break;
        }
        
        sequence_step_++;
    }

    // Member variables
    rclcpp::Publisher<trajectory_msgs::msg::JointTrajectory>::SharedPtr joint_trajectory_pub_;
    rclcpp::Publisher<std_msgs::msg::Float64MultiArray>::SharedPtr gripper_pub_;
    rclcpp::Subscription<sensor_msgs::msg::JointState>::SharedPtr joint_state_sub_;
    rclcpp::TimerBase::SharedPtr timer_;
    
    std::vector<std::string> joint_names_;
    std::vector<double> current_joint_positions_;
    int sequence_step_;
    bool gripper_closed_;
};

int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<PickPlaceController>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/pick_place_controller.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.duration import Duration
from sensor_msgs.msg import JointState
from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint
from std_msgs.msg import Float64MultiArray
import time

class PickPlaceController(Node):
    def __init__(self):
        super().__init__('pick_place_controller')
        
        # Publishers
        self.joint_trajectory_pub = self.create_publisher(
            JointTrajectory,
            '/joint_trajectory_controller/joint_trajectory',
            10
        )
        
        self.gripper_pub = self.create_publisher(
            Float64MultiArray,
            '/gripper_controller/commands',
            10
        )
        
        # Subscriber
        self.joint_state_sub = self.create_subscription(
            JointState,
            '/joint_states',
            self.joint_state_callback,
            10
        )
        
        # Initialize variables
        self.joint_names = ['joint1', 'joint2', 'joint3']
        self.current_joint_positions = [0.0, 0.0, 0.0]
        self.sequence_step = 0
        self.gripper_closed = False
        
        # Timer for pick and place sequence (every 3 seconds)
        self.timer = self.create_timer(3.0, self.execute_pick_place_sequence)
        
        self.get_logger().info('Pick and Place Controller (Python) initialized')
        
        # Define the pick and place sequence
        self.sequence = [
            {'name': 'Home Position', 'joints': [0.0, 0.0, 0.0], 'gripper': False},
            {'name': 'Approach Red Box', 'joints': [0.4, -0.3, 0.5], 'gripper': False},
            {'name': 'Lower to Pick', 'joints': [0.4, -0.1, 0.3], 'gripper': False},
            {'name': 'Close Gripper', 'joints': [0.4, -0.1, 0.3], 'gripper': True},
            {'name': 'Lift Object', 'joints': [0.4, -0.3, 0.5], 'gripper': True},
            {'name': 'Move to Target', 'joints': [-0.8, -0.3, 0.5], 'gripper': True},
            {'name': 'Lower to Place', 'joints': [-0.8, -0.1, 0.3], 'gripper': True},
            {'name': 'Open Gripper', 'joints': [-0.8, -0.1, 0.3], 'gripper': False},
            {'name': 'Retract', 'joints': [-0.8, -0.3, 0.5], 'gripper': False},
            {'name': 'Return Home', 'joints': [0.0, 0.0, 0.0], 'gripper': False}
        ]
    
    def joint_state_callback(self, msg):
        """Update current joint positions from joint state feedback"""
        for i, name in enumerate(msg.name):
            if name in self.joint_names:
                joint_index = self.joint_names.index(name)
                self.current_joint_positions[joint_index] = msg.position[i]
    
    def move_to_joint_positions(self, target_positions, duration=2.0):
        """Send joint trajectory command to move robot arm"""
        trajectory_msg = JointTrajectory()
        trajectory_msg.header.stamp = self.get_clock().now().to_msg()
        trajectory_msg.joint_names = self.joint_names
        
        # Create trajectory point
        point = JointTrajectoryPoint()
        point.positions = target_positions
        point.time_from_start = Duration(seconds=duration).to_msg()
        
        trajectory_msg.points = [point]
        self.joint_trajectory_pub.publish(trajectory_msg)
        
        self.get_logger().info(
            f'Moving to joint positions: [{target_positions[0]:.2f}, '
            f'{target_positions[1]:.2f}, {target_positions[2]:.2f}]'
        )
    
    def control_gripper(self, close):
        """Control gripper open/close"""
        gripper_msg = Float64MultiArray()
        if close:
            gripper_msg.data = [0.04, 0.04]  # Close gripper
            self.get_logger().info('🤏 Closing gripper')
        else:
            gripper_msg.data = [0.0, 0.0]   # Open gripper
            self.get_logger().info('✋ Opening gripper')
        
        self.gripper_pub.publish(gripper_msg)
        self.gripper_closed = close
    
    def execute_pick_place_sequence(self):
        """Execute the pick and place sequence step by step"""
        if self.sequence_step >= len(self.sequence):
            self.sequence_step = 0  # Reset sequence
            self.get_logger().info('🔄 Pick and place sequence completed! Restarting...')
            return
        
        current_step = self.sequence[self.sequence_step]
        step_num = self.sequence_step + 1
        
        self.get_logger().info(
            f'📍 Step {step_num}/10: {current_step["name"]}'
        )
        
        # Move joints
        self.move_to_joint_positions(current_step['joints'])
        
        # Control gripper if needed
        if current_step['gripper'] != self.gripper_closed:
            self.control_gripper(current_step['gripper'])
        
        self.sequence_step += 1

def main(args=None):
    rclpy.init(args=args)
    
    try:
        controller = PickPlaceController()
        rclpy.spin(controller)
    except KeyboardInterrupt:
        print('\n🛑 Pick and Place Controller stopped by user')
    finally:
        if rclpy.ok():
            rclpy.shutdown()

if __name__ == '__main__':
    main()